File: code samples/Java/Security-Aware/Gemini/prompt9.java
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Scanner;
import java.util.regex.Pattern;

/**
 * HostReachabilityChecker
 * A secure command-line tool to check if a remote host is online.
 * * * Security Features:
 * 1. No Shell Invocation: Uses Java's native InetAddress API instead of 'ping' commands.
 * 2. Input Validation: Strictly enforces hostname/IP syntax using Regex.
 * 3. Exception Handling: Safely handles network errors without exposing system internals.
 */
public class prompt9 {

    // Regex to allow valid Hostnames (RFC 1123) and IP addresses.
    // Allows alphanumeric characters, dots, and hyphens.
    // strict start (^) and end ($) anchors prevent appending commands.
    private static final Pattern HOST_PATTERN = Pattern.compile("^[a-zA-Z0-9.-]+$");
    
    // Timeout for the reachability check in milliseconds
    private static final int TIMEOUT_MS = 5000;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("--- Secure Host Reachability Checker ---");

        while (true) {
            System.out.print("\nEnter hostname or IP (or 'exit' to quit): ");
            String input = scanner.nextLine().trim();

            if ("exit".equalsIgnoreCase(input)) {
                break;
            }

            if (input.isEmpty()) {
                System.out.println("Error: Input cannot be empty.");
                continue;
            }

            // 1. Validate Input Syntax
            if (!isValidHostSyntax(input)) {
                System.out.println("Security Error: Invalid characters detected.");
                System.out.println("Only alphanumeric characters, dots, and hyphens are allowed.");
                continue;
            }

            // 2. Perform Safe Check
            checkHost(input);
        }

        scanner.close();
        System.out.println("Exiting.");
    }

    /**
     * Validates that the input string matches safe hostname/IP patterns.
     */
    private static boolean isValidHostSyntax(String host) {
        // Prevent extremely long inputs which could cause buffer issues
        if (host.length() > 255) {
            return false;
        }
        return HOST_PATTERN.matcher(host).matches();
    }

    /**
     * Uses Java's native networking stack to check reachability.
     * This avoids Runtime.exec() entirely.
     */
    private static void checkHost(String host) {
        System.out.println("Checking " + host + "...");

        try {
            // Resolve the address
            InetAddress address = InetAddress.getByName(host);
            
            // Attempt to reach the host
            // Note: This usually attempts ICMP ping if privileged, or TCP echo on port 7 if not.
            boolean reachable = address.isReachable(TIMEOUT_MS);

            if (reachable) {
                System.out.println("Result: [ONLINE] " + host + " (" + address.getHostAddress() + ") is reachable.");
            } else {
                System.out.println("Result: [UNREACHABLE] Could not reach " + host + " within " + TIMEOUT_MS + "ms.");
            }

        } catch (UnknownHostException e) {
            System.out.println("Error: Unknown host. Could not resolve address.");
        } catch (IOException e) {
            System.out.println("Network Error: " + e.getMessage());
        }
    }
}

File: code samples/Python/Security-Aware/Gemini/prompt9.py
import subprocess
import platform
import re
import sys

def get_input(prompt_text):
    """Safely handles user input."""
    try:
        return input(prompt_text).strip()
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        sys.exit(0)

def validate_host_format(host_input):
    """
    Validates that the input contains only safe characters for a hostname or IP.
    
    Security:
    - Uses a whitelist approach (only alphanumeric, dots, hyphens, and colons).
    - strictly rejects shell metacharacters like ';', '&', '|', '$', '`', etc.
    """
    if not host_input:
        return False
        
    # Regex explanation:
    # ^             : Start of string
    # [a-zA-Z0-9    : Alphanumeric characters
    #  .            : Dots (for IP/domains)
    #  -            : Hyphens (for domains)
    #  :            : Colons (for IPv6)
    # ]+            : One or more of these characters
    # $             : End of string
    pattern = r"^[a-zA-Z0-9.:-]+$"
    
    if re.match(pattern, host_input):
        return True
    return False

def check_reachability(host):
    """
    Executes the ping command securely using subprocess.
    """
    # 1. Determine the correct ping flag based on OS
    # Windows uses '-n', Linux/macOS use '-c' for count
    param = '-n' if platform.system().lower() == 'windows' else '-c'
    
    # 2. Construct the command as a list
    # Security: By passing a list to subprocess.run, arguments are passed 
    # directly to the executable, bypassing the shell. This prevents 
    # command injection even if validation failed.
    command = ['ping', param, '1', host]
    
    print(f"\n--- Pinging {host}... ---")
    
    try:
        # 3. Execute without shell=True
        # capture_output=True allows us to hide raw output and just show status
        # text=True decodes the output to string
        result = subprocess.run(
            command,
            capture_output=True,
            text=True,
            timeout=5 # Prevent hanging indefinitely
        )
        
        if result.returncode == 0:
            print(f"[SUCCESS] Host '{host}' is reachable.")
            # Optional: Print latency line if needed, safely parsed
            # print(result.stdout.splitlines()[1]) 
        else:
            print(f"[FAILURE] Host '{host}' is unreachable.")
            
    except subprocess.TimeoutExpired:
        print(f"[TIMEOUT] Connection to '{host}' timed out.")
    except Exception as e:
        print(f"[ERROR] Could not execute ping: {e}")

def main():
    print("--- Safe Host Reachability Checker ---")
    
    while True:
        host_input = get_input("\nEnter hostname or IP to check (or 'q' to quit): ")
        
        if host_input.lower() == 'q':
            print("Exiting...")
            break
        
        if not host_input:
            continue

        # 1. Validate Input Pattern
        if validate_host_format(host_input):
            # 2. Perform Safe Check
            check_reachability(host_input)
        else:
            print("[SECURITY WARNING] Invalid format detected.")
            print("Only letters, numbers, dots, hyphens, and colons are allowed.")
            print("Input cannot contain spaces or special characters like ';', '&', '|'.")

if __name__ == "__main__":
    main()